### ISSUE TITLE
Decide how mixture-of-experts routing is typed under static shapes

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#102]
> Filed against the `rfc.yml` template: this asks what a mixture-of-experts layer's type signature is, which is a core-invariant question rather than a feature request.

## Summary

A mixture-of-experts layer is dynamically shaped in the one place a statically shaped tensor library cannot follow: the number of tokens each expert receives is a function of the input. `Tensor<S, B, K, G>` puts the shape in the type, and there is no `S` that means "however many tokens the router sent here". This RFC picks how that is expressed, because the choice determines the layer's public signature and cannot be changed later without breaking every model built on it.

### The shape problem, concretely

Top-`k` routing over `E` experts:

```text
tokens        [T, d_model]          T static
router logits [T, E]                static, fine
top-k indices [T, k]                static, fine
expert e gets [n_e, d_model]        n_e is DATA-DEPENDENT.  sum(n_e) = T * k
expert output [n_e, d_ff] -> [n_e, d_model]
combine back  [T, d_model]          static again
```

Everything is static except the span in the middle, and that span is where all the compute is. `n_e` varies per step, per expert, and per input batch. Nothing in the shape algebra can name it.

### Why this is not the primitives issue

#103 adds `one_hot`, `sort`, `nonzero`, `bincount`, `scatter_add`, `repeat_interleave`, `log_softmax`, and `logsumexp` to the catalog. Those are necessary and not sufficient: with all of them present, the question of what `MoE::forward` *returns* and what bounds it carries is still open. That question is this issue.

### Why it interacts with #100

Routing produces a `[T, k]` index tensor applied to `[T, d_model]` activations, and a `[T, E]` gate tensor multiplied into `[T, d_model]` outputs. Both are shape pairs the reflexive `ShapeEq` bound cannot express. Whatever #100 decides is the mechanism this builds on, which is why that decision is sequenced first.

## Invariant and soundness impact

Touches **Shape**, and touches the autograd tape indirectly.

- **Shape.** Options A and C keep every user-visible shape static; option B introduces a `Dyn` span inside a layer whose inputs and outputs are static. That is already an expressible mixture in-tree, so this is a scope question rather than a new capability.
- **Autograd.** Top-`k` selection is not differentiable in the routing decision. The gradient flows through the *gate weights*, not through which expert was chosen. That contract has to be written down, not left to whatever the tape happens to record. A layer that silently produces no gradient for its router is a correctness defect that looks like a convergence problem.
- **Determinism.** The combine step accumulates `k` contributions per token. In floating point that is order-dependent, so `scatter_add`'s determinism guarantee (#103) is part of this contract rather than adjacent to it.

No `unsafe`, no FFI, no device or dtype contract change.

## Detailed design

Three candidates for the dynamic span.

### A. Capacity-factor padding

Fix each expert's operand at `[capacity, d_model]` where `capacity = ceil(T / E) * factor`. Pad underfull experts, drop overflow.

```rust
// proposed
#[module]
struct MoeFeedForward<const E: usize, S: Shape, B: Backend> {
    router: Linear<S, B>,
    experts: [FeedForward<S, B>; E],
}

// capacity is a const parameter, so every expert operand is a static shape
type ExpertInput = Tensor<s![CAPACITY, 512], Backend, f32, Grad>;
```

Every shape stays static and the type system keeps working exactly as designed. The costs are real and non-negotiable: tokens are **dropped** at overflow, compute is spent on padding, and the capacity factor becomes a hyperparameter the user must tune. This is what the TPU-era implementations did, and it is the only candidate that preserves static shapes end to end.

### B. A `Dyn` boundary at dispatch

Route into `Tensor<Dyn, B, K, G>` per expert, run the expert, return to the static shape at combine.

```rust
// proposed
fn forward(&self, x: Tensor<s![T, D], B, f32, Grad>) -> Result<Tensor<s![T, D], B, f32, Grad>> {
    let gates = self.router.forward(x.clone())?.softmax(1)?;   // [T, E]
    let (weights, indices) = gates.topk(K_EXPERTS, 1)?;        // [T, k] each
    let mut out = Tensor::<s![T, D], B, f32, Grad>::zeros(())?;
    for (e, expert) in self.experts.iter().enumerate() {
        let rows = indices.eq_scalar(e as i64)?.nonzero()?;    // Dyn: [n_e, 1]
        let tokens = x.index_select(&rows, 0)?.into_dyn();     // Dyn: [n_e, D]
        let y = expert.forward(tokens)?;                       // Dyn
        out = out.scatter_add(&rows, &y.try_static::<s![T, D]>()?, 0)?;
    }
    Ok(out)
}
```

No dropped tokens, no wasted compute. The static proof is absent for the span where the shapes are dynamic, and the per-expert loop is `E` separate launches. Smallest new mechanism: `Dyn`, mixed shapes, and the `Dyn`-to-static boundary all exist today.

### C. Grouped execution over a sorted buffer

Sort tokens by assigned expert, keep one `[T * k, d_model]` buffer, run a **grouped matmul** with per-expert row offsets.

```text
permutation  argsort(expert_id)        -> [T*k]
offsets      cumsum(bincount(ids, E))  -> [E + 1]           static extent E+1
buffer       gather(x, permutation)    -> [T*k, d_model]    static
grouped_gemm(buffer, weights, offsets) -> [T*k, d_ff]       static
combine      scatter_add(inverse perm) -> [T, d_model]      static
```

The outer shape is static, nothing is dropped, nothing is padded, and the dynamic part becomes an **offset array rather than a shape**. This is the key insight: `n_e` never needs to be a type because it is never a tensor extent, only a row range inside a statically shaped buffer. It needs a grouped GEMM, which the current CUDA backend does not have (#103 adds it, on top of the batched GEMM in #85), and it is how current high-performance implementations work.

### Recommendation

**Specify the public signature so that B and C are the same type**, and only the execution strategy differs. B is the honest interim that runs on CPU today; C is where this ends up once grouped GEMM exists. A is rejected for the default because dropping tokens is a silent accuracy loss, though it should remain available as an explicit configuration for users who want bounded per-expert compute.

### Also to be settled here

Each of these changes the signature, so none can be deferred:

1. **Are `E` and `k` type-level or runtime?** `E` should be const (it determines the parameter count and therefore the state layout). `k` is arguably runtime, but a const `k` lets the `[T, k]` intermediate be statically shaped, which is worth more than the flexibility.
2. **Where does the load-balancing auxiliary loss live?** It is a loss computed *inside* a layer, which `Module` has no notion of. Options: return it in the layer's `Output` associated type as a tuple, accumulate it into a trainer-visible side channel, or expose it as a method the caller must remember to add. The first is the only one that cannot be silently forgotten.
3. **What is the gradient contract through top-`k`?** Straight-through, gate-weight-only, or documented as non-differentiable in the selection.

## Drawbacks and alternatives

**Declaring MoE out of scope** is defensible: it is one architecture family, and #103's primitives would let a determined user build it by hand. It would also mean the type system's central claim, that the shape is part of the type, has an architecture-shaped hole the library never addresses, discovered by each user independently. Answering it deliberately, even with "use `Dyn` here and here is why", is worth more than leaving it open.

**Deferring only the auxiliary-loss question** is not viable. It changes `Module::Output`, so it is part of the signature or it is a breaking change later.

## Acceptance criteria

- The decision is recorded in `PROPOSALS.md` with the rejected alternatives, per the repository's design-ruling convention.
- The MoE layer's public type signature is specified and does not change between the interim and target execution strategies.
- The auxiliary-loss and gradient contracts are documented in `docs/OPERATION_SEMANTICS.md`.
- A prototype layer under `incin::experimental` demonstrates the chosen signature on CPU, even if its execution strategy is the slow one.

## How this interoperates

- **#100** decides the mechanism for every shape pair here. Sequenced first, and this issue should not be resolved before it.
- **#103** supplies the eight missing catalog operations plus `grouped_matmul`. Independent of this decision except for `nonzero`, whose data-dependent output extent is the clearest instance of the question this RFC answers, so the two must agree.
- **#87 and #88** bring the existing routing primitives (`topk`, `cumsum`, `argsort`, `gather`, `scatter`, `index_select`) to CUDA. Option C is unimplementable on GPU without them.
- **#85** provides the batched GEMM that `grouped_matmul` extends with per-expert offsets. Option C's performance argument rests on it.
- **#101** shares the `Module` surface: an MoE feed-forward block substitutes for the dense one inside a transformer layer, so its `Output` associated type must compose with `TransformerDecoderLayer`. That is the practical constraint behind the auxiliary-loss question.
- **`Module::Output` and state traversal.** Whichever option is chosen, `E` experts must appear in the state snapshot with stable `StatePath`s, so a checkpoint written with one expert count is refused rather than partially loaded against another.
- **#99 (distributed)** is where MoE eventually becomes interesting at scale: expert parallelism is a placement over the existing mesh axes. The signature chosen here should not foreclose sharding experts across devices, even though that is out of 0.2.0 scope.
- **`incin::experimental`** is the right home for the prototype layer, matching how compiled execution and the trainer are staged.

## Additional context

Relevant code: `crates/incin-core/src/shapes/shape.rs` (`Dyn`, `DynShape`, `DimCons`), `crates/incin-core/src/tensor/ops/index.rs` (`ShapeEq`), `crates/incin-core/src/nn/module.rs` (`Module::Output`), `PROPOSALS.md`.

Depends on #100. Primitives in #103; grouped GEMM there depends on #85. Blocks the MoE implementation work.
